### PR TITLE
Make keyboard interactions in the settings menu more pleasant

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
#78868 improved the keyboard interactions with the settings page. This PR goes a bit further by allowing more than just "space" to toggle the checkboxes.

r? @jyn514

### CODE REVIEW

File: src/librustdoc/html/static/main.js [hunk 1]
```diff
@@ -40,6 +40,29 @@ if (!DOMTokenList.prototype.remove) {
     };
 }
 
+
+// Gets the human-readable string for the virtual-key code of the
+// given KeyboardEvent, ev.
+//
+// This function is meant as a polyfill for KeyboardEvent#key,
+// since it is not supported in Trident.  We also test for
```
Comment (jyn514) [inline src/librustdoc/html/static/main.js:48]: It turns out Trident is IE 11. But Android doesn't have it either and has a ton more market-share, so I think we should say that instead.
```suggestion
// since it is not supported in IE 11 or Chrome for Android. We also test for
```
https://caniuse.com/?search=KeyboardEvent

File: src/librustdoc/html/static/settings.js [hunk 2]
```diff
@@ -14,10 +14,24 @@
...
+                break;
+        }
+    }
+
     function setEvents() {
         var elems = {
-            toggles: document.getElementsByClassName("slider"),
-            selects: document.getElementsByClassName("select-wrapper")
+            toggles: Array.prototype.slice.call(document.getElementsByClassName("slider")),
+            selects: Array.prototype.slice.call(document.getElementsByClassName("select-wrapper")),
```
Comment (jyn514) [inline src/librustdoc/html/static/settings.js:34]: What does this do?

File: src/librustdoc/html/static/settings.js [hunk 3]
```diff
@@ -50,5 +66,5 @@
         }
     }
 
-    setEvents();
+    window.addEventListener("DOMContentLoaded", setEvents);
```
Comment (jyn514) [inline src/librustdoc/html/static/settings.js:69]: Why this change?

File: src/librustdoc/html/static/settings.js [hunk 4]
```diff
@@ -32,6 +46,8 @@
                 toggle.onchange = function() {
                     changeSetting(this.id, this.checked);
                 };
+                toggle.onkeyup = handleKey;
+                toggle.onkeyrelease = handleKey;
```
Comment (jyn514) [inline src/librustdoc/html/static/settings.js:50]: Doesn't this mean the handler will be called multiple times?

**Review by jyn514 — COMMENTED** [review]
Is this waiting on https://github.com/rust-lang/rust/pull/78868?

**Comment by GuillaumeGomez** [comment]
> Is this waiting on #78868?

Not really. It'll possible to use it only when it's in but the changes don't conflict.

File: src/librustdoc/html/static/main.js (hunk 1 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:48]: Oh sure. Just moved the function so didn't look at the docs. Thanks!

File: src/librustdoc/html/static/settings.js (hunk 2 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/settings.js:34]: A very "funny" bug. Basically, it's only possible to view it on chrome. `document.getElementsByClassName` returns an iterator, but every time you try to access a value, on chrome it goes through all the elements and then returns the one you asked, which made rustdoc pages very slow to render on chrome (and which is why I created the `onEachLazy` function).

In this case, it seems like they wanted to do it this way, but I guess using `onEachLazy` would make more sense?

File: src/librustdoc/html/static/settings.js (hunk 3 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/settings.js:69]: I realized that the script was loaded before the DOM. In itself, it seems to be working but I guess it doesn't hurt to be sure that the DOM is actually there before trying to work on it.

File: src/librustdoc/html/static/settings.js (hunk 4 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/settings.js:50]: The event callback is set to a given element, so only once per element. If you use the key but don't have focus on it, nothing will happen.

File: src/librustdoc/html/static/settings.js (hunk 2 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/settings.js:34]: Ok, definitely better with `onEachLazy`. Sending the update soon.

**Comment by GuillaumeGomez** [comment]
Updated!

File: src/librustdoc/html/static/settings.js [hunk 5]
```diff
@@ -14,41 +14,46 @@
...
+    function handleKey(ev) {
+        // Don't interfere with browser shortcuts
+        if (ev.ctrlKey || ev.altKey || ev.metaKey) {
+            return;
+        }
+        switch (getVirtualKey(ev)) {
+            case "Enter":
+            case "Return":
+            case "Space":
+                ev.target.checked = !ev.target.checked;
```
Comment (notriddle) [inline src/librustdoc/html/static/settings.js:26]: If you’re going to trap an event in JS, you should also call `ev.preventDefault()`

**Review by notriddle — CHANGES_REQUESTED** [review]

File: src/librustdoc/html/static/settings.js (hunk 5 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/settings.js:26]: I was wondering if it made sense but I guess it does. I'll add it then, thanks!

**Comment by jyn514** [comment]
> Not really. It'll be possible to use it only when it's in but the changes don't conflict.

Do you mind rebasing this so it's possible to test? Right now I can't try it out.

**Comment by GuillaumeGomez** [comment]
Done! You can now test with all changes included.

**Comment by jyn514** [comment]
@bors r+

Works great, thanks!
